### PR TITLE
New batch of tuple tests ported to VB

### PR DESCRIPTION
**Pull request by jcouv**
I flagged a few areas of interest with TODO REVIEW.

Fixed two small code issues that were uncovered.

@dotnet/roslyn-compiler for review.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 1]
```diff
@@ -7673,36 +7673,6 @@ static void Test<T>( (Func<(int alice, int bob), T> f1, Func<(int, int), T> f2)
 ");
         }
 
-        [WorkItem(10801, "https://github.com/dotnet/roslyn/issues/10801")]
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/10801")]
-        public void Inference06()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: The issue was closed based on LDM decision.

**Comment by jcouv** [comment]
>             ' TODO REVIEW In C#, the third one is selected

Are overload resolution rules expected to be different in VB?

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10443 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by jcouv** [comment]
>         Public Sub Inference13()

I'm investigating this one. Running the compiled code produces a cast exception.

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10746 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by jcouv** [comment]
>         Public Sub TupleWithoutFeatureFlag()

@VSadov Were tuples in VB added for a specific language version? Is that planned?

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11039 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by jcouv** [comment]
>             'Assert.Null(m1Tuple.ComImportCoClass) ' TODO REVIEW

This API doesn't exist in VB. Should I replace it with some other VB-specific checks?

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11292 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by cston** [comment]
>             Assert.Equal({}, symbols.Skip(common).Select(Function(s) s.ToTestDisplayString()).ToArray())

Is this assert redundant given the assert below?

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11503 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by cston** [comment]
>             'Assert.Null(m3Item8.GetUseSiteDiagnostic())

Is this assert failing. If so, consider adding an issue #.

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11601 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by cston** [comment]
>             'Assert.Null(m3Item8.GetUseSiteDiagnostic())

`m3Item8` is not defined. Same comment for tests below.

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11763 in 36101e9. [](commit_id = 36101e935f73784b8f1cda4a010525dd17871243, deletion_comment = False)

**Comment by cston** [comment]
LGTM

**Comment by jcouv** [comment]
>             Assert.Equal({}, symbols.Skip(common).Select(Function(s) s.ToTestDisplayString()).ToArray())

Yes, I'll remove the assert below. The first one provides better output when the test fails.

---
In reply to: [250288878](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250288878) [](ancestors = 250288878)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11503 in 3590fd8. [](commit_id = 3590fd81dc0998b888e877855a6dbeeb13ab3491, deletion_comment = False)

**Comment by jcouv** [comment]
>             Assert.Equal({}, symbols.Skip(common).Select(Function(s) s.ToTestDisplayString()).ToArray())

On second thought, no. When `symbols` is shorter than `baseLine`, the first assert will not fail. I need to keep both (the first one is useful when writing new tests).

---
In reply to: [250293509](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250293509) [](ancestors = 250293509,250288878)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11503 in 3590fd8. [](commit_id = 3590fd81dc0998b888e877855a6dbeeb13ab3491, deletion_comment = False)

**Comment by jcouv** [comment]
>             'Assert.Null(m3Item8.GetUseSiteDiagnostic())

The issue here is that there is no `GetUseSiteDiagnostic` in VB. I'm going to replace with calling `GetUseSiteErrorInfo`.

---
In reply to: [250289399](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250289399) [](ancestors = 250289399)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11601 in 3590fd8. [](commit_id = 3590fd81dc0998b888e877855a6dbeeb13ab3491, deletion_comment = False)

**Comment by cston** [comment]
>             Assert.Equal({}, symbols.Skip(common).Select(Function(s) s.ToTestDisplayString()).ToArray())

In that case, there is a helper method that handles the entire comparison: `AssertEx.Equal(symbols, baseLine, itemInspector:=ToTestDisplayString)`

---
In reply to: [250294062](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250294062) [](ancestors = 250294062,250293509,250288878)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11503 in 3590fd8. [](commit_id = 3590fd81dc0998b888e877855a6dbeeb13ab3491, deletion_comment = False)

**Comment by jcouv** [comment]
>             'Assert.Null(m3Item8.GetUseSiteDiagnostic())

Fixed


---
In reply to: [250290081](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250290081) [](ancestors = 250290081)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11763 in 3590fd8. [](commit_id = 3590fd81dc0998b888e877855a6dbeeb13ab3491, deletion_comment = False)

**Comment by jcouv** [comment]
>             Assert.Equal({}, symbols.Skip(common).Select(Function(s) s.ToTestDisplayString()).ToArray())

Very handy. Thanks

---
In reply to: [250296385](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250296385) [](ancestors = 250296385,250294062,250293509,250288878)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11503 in 3590fd8. [](commit_id = 3590fd81dc0998b888e877855a6dbeeb13ab3491, deletion_comment = False)

**Comment by jcouv** [comment]
@VSadov Please take a look, especially the remaining REVIEW markers. Should I file issues on those?

**Comment by VSadov** [comment]
>             ' TODO REVIEW In C#, the third one is selected

I think so. Second is not a candidate in C# at all since lambdas do not have default. In VB they have anonymous delegate types.
Once you have two candidates in the set  the one with objects wins. I think because "least generic" rule.

If instead of "Function" you use "AddressOf < some method returning int >", it would be more like in C# since AddressOf is truly typeless.

---
In reply to: [250268704](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250268704) [](ancestors = 250268704)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10443 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by VSadov** [comment]
>             ' TODO REVIEW In C#, the third one is selected

I think same explanation as in the other case.

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10479 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by VSadov** [comment]
>             ' TODO REVIEW

I am a bit surprised that C# case is a failure. Would we not just drop the conflicting names in this scenario?

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10710 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by VSadov** [comment]
>             ' TODO REVIEW In C#, there are errors because names matter during best type inference

Same as the other one. It seems that name dropping would happen, but maybe I am not noticing another reason for the failure in C#.

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10806 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by VSadov** [comment]
>         Public Sub TupleWithoutFeatureFlag()

Yes!  If we do not condition it on VB15, we have a bug here.

---
In reply to: [250269681](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250269681) [](ancestors = 250269681)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11039 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by VSadov** [comment]
>         Public Sub TupleWithoutFeatureFlag()

Thanks for checking this!!

---
In reply to: [250879957](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250879957) [](ancestors = 250879957,250269681)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11039 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by VSadov** [comment]
>             ' TODO REVIEW

It seems like tuple element names affect overload resolution while not being used at all, and that seems wrong. It looks like we have a bug in C#.

---
In reply to: [250879604](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250879604) [](ancestors = 250879604)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10710 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by jcouv** [comment]
>             ' TODO REVIEW

Filed [https://github.com/dotnet/roslyn/issues/14246](https://github.com/dotnet/roslyn/issues/14246) (VB) and [https://github.com/dotnet/roslyn/issues/14247](https://github.com/dotnet/roslyn/issues/14247) (C#) for `Inference11` test.

---
In reply to: [250916290](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250916290) [](ancestors = 250916290,250879604)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10710 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by jcouv** [comment]
>             ' TODO REVIEW In C#, the third one is selected

Filed issue [https://github.com/dotnet/roslyn/issues/14255](https://github.com/dotnet/roslyn/issues/14255) to track the difference in overload resolution when parameters are flat versus in a tuple.

---
In reply to: [250878941](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250878941) [](ancestors = 250878941,250268704)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:10443 in 6d3a7d9. [](commit_id = 6d3a7d94d4501ee5529a1f8b8e1141c2660ee18d, deletion_comment = False)

**Comment by jcouv** [comment]
@VSadov The latest commit addresses the PR feedback (mostly adding tests and filing new issues).
Also, I found the VB15 flag, so that one is fixed.

**Comment by jcouv** [comment]
@dotnet-bot test windows_release_unit32_prtest please

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 2]
```diff
@@ -7936,11 +7974,13 @@ static void Test3<T>(ref T x, ref T y)
...
-                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test3").WithArguments("C.Test3<T>(ref T, ref T)").WithLocation(11, 9)
-                );
+            comp.VerifyDiagnostics();
+
+            // This diagnostic should not be there. The inference should succeed as Test3<(int, int)>
+            // (11,9): error CS0411: The type arguments for method 'C.Test3<T>(ref T, ref T)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+            //         Test3(ref ab, ref cd);
+            //Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test3").WithArguments("C.Test3<T>(ref T, ref T)").WithLocation(11, 9)
+            // See https://github.com/dotnet/roslyn/issues/14247"
         }
```
Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: Thanks!

**Comment by VSadov** [comment]
>         Public Sub TupleWithoutFeatureFlag()

Is there a tracking bug for this? We should check language version.

---
In reply to: [250880003](https://github.com/dotnet/roslyn/pull/14149#issuecomment-250880003) [](ancestors = 250880003,250879957,250269681)

---
Refers to: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb:11039 in 3911967. [](commit_id = 39119677f090b1726e087541bd437b3473c3a8bb, deletion_comment = False)

**Comment by VSadov** [comment]
:shipit:

**Comment by jcouv** [comment]
> Regarding TupleWithoutFeatureFlag
It was a test bug. I found the language version VB14 and the test is correct and passing (it reports an error that tuples require VB15).

**Pull request merged**